src/tcache.c

```diff
@@ -516,15 +518,22 @@ tcache_alloc_small_hard(tsdn_t *tsdn, arena_t *arena,
 
 	assert(tcache_slow->arena != NULL);
 	assert(!tcache_bin_disabled(binind, cache_bin, tcache_slow));
+
+	uint8_t lg_div = tcache_slow->lg_fill_div[binind];
+	uint8_t lg_offset = tcache_slow->lg_fill_div_offset[binind];
+	assert(lg_offset < lg_div);
 	cache_bin_sz_t nfill = cache_bin_ncached_max_get(cache_bin)
-	    >> tcache_slow->lg_fill_div[binind];
+	    >> (lg_div - (opt_experimental_tcache_gc ? lg_offset : 0));
 	if (nfill == 0) {
 		nfill = 1;
 	}
 	arena_cache_bin_fill_small(tsdn, arena, cache_bin, binind,
 	    /* nfill_min */ opt_experimental_tcache_gc ?
 	    ((nfill >> 1) + 1) : nfill, /* nfill_max */ nfill);
 	tcache_slow->bin_refilled[binind] = true;
+	if (lg_offset + 1 < lg_div) {
+		tcache_slow->lg_fill_div_offset[binind]++;
```

Comment:
I think it's worth adding a comment here to explain how `lg_fill_div_offset` is updated and the reason why we don't simply reduce one to `lg_fill_div`.